datafiles/templates/upload.html.st

```diff
@@ -72,37 +66,32 @@ See the notes at the bottom of the page.
   </li>
 
   <li>
-    Categories are determined by whatever you put in the <code>Category</code> field
-    (there's no agreed list of category names yet).
+    Categories are determined by whatever you put in the <code>Category</code> field.
+    You should try to pick existing categories when possible.
     You can have more than one category, separated by commas. If no other versions of
     the package exist, the categories automatically become the package's tags.
   </li>
 
   <li>
-    Documentation for library packages should be generated by a maintainer.
-    The means of doing this is still up in the air.
+    Occasional changes to the GHC base package can mean that some work needs to be done to make packages compatible across a range of versions.
+    See <a href="https://github.com/haskell-infra/hackage-trustees/blob/master/cookbook.md">these notes</a> for some tips in how to do so.
+    
+    There are some notes for upgrading <a href="http://www.haskell.org/haskellwiki/Upgrading_packages">much older</a> packages as well.
   </li>
 
-  <li>
-    We have moved to Haddock 2, and expect some glitches.
+<h3>Documentation</h3>
 
-    If you notice anything broken, please report it on the
-    <a href="http://trac.haskell.org/haddock">Haddock bug tracker</a>.
-  </li>
+<p>
+The hackage-server attempts to build documentation for library packages, but this can fail. Maintainers can generate their own documentation and upload it by using something along the lines of the shell script below (note that the last two commands are the key ones):</p>
+<pre>
+set -e
 
-  <li>
-    In GHC 6.8, several modules were split from the <code>base</code> package
-    into other packages.
+dir=$(mktemp -d dist-docs.XXXXXX)
+trap 'rm -r "$dir"' EXIT
 
-    See <a href="http://www.haskell.org/haskellwiki/Upgrading_packages">these notes</a>
-    on making packages work with a range of versions of GHC.
-  </li>
-
-  <li>
-    While <a href="http://www.haskell.org/haddock/">Haddock 2</a>
-    accepts GHC features, it is also more picky about comment syntax than
-    the old version.
-  </li>
+cabal haddock --builddir="$dir" --for-hackage --haddock-option=--hyperlinked-source
+cabal upload -d $dir/*-docs.tar.gz
```

Comment:
Note that starting with cabal 2.0 one also needs to use the `--publish` flag to `cabal upload` (as it defaults to candidates)